grudge/trace_pair.py

```diff
@@ -272,61 +281,144 @@ def connected_ranks(dcoll: DiscretizationCollection):
     return get_connected_partitions(dcoll._volume_discr.mesh)
 
 
+def _flatten_to_numpy_from_container(
+        actx: ArrayContext,
+        ary: ArrayOrContainerT) -> Tuple[ArrayT, Optional[Dict[Any, Any]]]:
+    """Flattens an arbitrary array container *ary* into a one-dimensional
+    numpy array suitable for communication via mpi4py.
+
+    :returns: A tuple containing the flattened result and an OrderedDict
+        providing a template for reconstructing data from the flattened result.
+        If *ary* is not a container, then *None* is returned for the template.
+    """
+    if is_array_container(ary):
+        # Create a dictionary template (order is important) of keys/values
+        # to remember how data is organized in the flattened numpy array
+        data_template = OrderedDict()
+        flat_data = []
+        # NOTE: The ordering of data in the flattened array will be determined
+        # by the ordering of key/value pairs returned by *serialize_container*
+        # and "remembered" by *data_template*
+        for key, value in serialize_container(ary):
+            # NOTE: serialize_container does not serialize a
+            # nested object array inside an array container.
+            if isinstance(value, np.ndarray) and value.dtype == "O":
+                flat_data.extend(to_numpy(value, actx))
+            else:
+                flat_data.append(to_numpy(value, actx))
+            # NOTE: if the value associated with *key* was originally an
+            # object array, we want to remember this during reconstruction
+            data_template[key] = value
+        return np.concatenate(flat_data), data_template
+    else:
+        # If *ary* is not a container-type, then *data_template* is left empty
+        # since there is only one 'component'
+        return flatten_to_numpy(actx, ary), None
+
+
+def _unflatten_from_numpy_to_container(
+        actx: ArrayContext,
+        container_template: ArrayOrContainerT,
+        data_template: Optional[Dict[Any, Any]],
+        numpy_ary, discr) -> ArrayOrContainerT:
```

Comment:
I'm confused why you need both a `container_template` and a `data_template` here? Wouldn't the `container_template` already prescribe how the data is laid out?

Reply:
Solved thanks to this :+1: https://github.com/inducer/arraycontext/pull/91